File: 06/Assembler/src/main/java/Parser.java
//package ...;
import java.io.*;

public class Parser {
    private BufferedReader bufferFile;
    private String currentLine;

    /**
     *Opens the input file/stream and gets ready to parse it.
     *Arguments: Input file or stream
     */
    public Parser(File file) throws IOException {
        Reader reader = new FileReader(file);
        bufferFile = new BufferedReader(reader);
        currentLine = null;
    }

    /**
     *Are there more lines in the input?
     */
    public boolean hasMoreLines() throws IOException {
        currentLine=bufferFile.readLine();
        return currentLine != null;
    }

    /**
     *Skips over whitespace and comments, if necessary.
     * Reads the next instruction from the input, and makes it the current instruction.
     * This method should be called only if hasMoreLines is true.
     * Initially there is no current instruction
     */
    public void advance () throws IOException{
        // Assuming currentLine already contains the next line from hasMoreLines
        // Trim the current line to remove leading and trailing whitespace
        currentLine = currentLine.trim();

        // If the line is a comment or empty, keep reading the next line
        while (currentLine.isEmpty()) {
            currentLine = bufferFile.readLine();
            if (currentLine == null) { // End of file reached
                break;
            }
            currentLine = currentLine.trim();
        }
    }

    /**
     *Returns the type of the current instruction:
     * A_INSTRUCTION for @xxx,where xxx is either a decimal number or a symbol.
     * (constants) C_INSTRUCTION for dest=comp; jump
     * L_INSTRUCTION for (xxx), where xxx is a symbol.
     */
    public String instructionType (){
        if (currentLine.startsWith("@")) {
            return "A_INSTRUCTION";
        } else if (currentLine.startsWith("(") && currentLine.endsWith(")")) {
            return "L_INSTRUCTION";
        } else {
            return "C_INSTRUCTION";
        }
    }

    /**
     *If the current instruction is (xxx), returns the symbol xxx.
     *If the current instruction is @xxx, returns the symbol or decimal xxx (as a string).
     * Should be called only if instructionType is A_INSTRUCTION or L_INSTRUCTION.
     */
    public String symbol (){
        if (currentLine.startsWith("@")) {
            return currentLine.substring(1);
        }
        // Assuming the method called only for A_INSTRUCTION or L_INSTRUCTION
        // L_INSTRUCTION : (XXXX) so I want to return substring(1,5)=substring(1,length-1). the length is 6
        else return currentLine.substring(1,currentLine.length()-1);
    }

    /**
     *Returns the symbolic dest part of the current C-instruction (8 possibilities).
     *Should be called only if instructionType is C_INSTRUCTION.
     */
    public String dest (){
        int posFirstSep = currentLine.indexOf('=');
        // dest does not exist
        if(posFirstSep == -1)
            return null;
        // there is a dest part
        return currentLine.substring(0, posFirstSep);
    }

    /**
     *Returns the symbolic comp part of the current C-instruction (28 possibilities).
     *Should be called only if instructionType is C_INSTRUCTION.
     */
    public String comp (){
        int posFirstSep = currentLine.indexOf('=');
        int posSecondSep = currentLine.indexOf(';');

        // dest does not exist so comp must be first
        if(posFirstSep == -1)
            return  currentLine.substring(0, posSecondSep);

        // dest exists and jump does not
        if (posSecondSep == -1)
            return  currentLine.substring(posFirstSep+1 );

        // dest exists and jump exists
        return  currentLine.substring(posFirstSep+1 , posSecondSep);
    }

    /**
     *Returns the symbolic jump part of the current C-instruction (8 possibilities).
     *Should be called only if instructionType is C_INSTRUCTION.
     */
    public String jump (){
        int posSecondSep = currentLine.indexOf(';');
        // jump does not exist
        if(posSecondSep == -1)
            return null;
        // there is a jump part
        return currentLine.substring(posSecondSep+1);
    }

}

File: software/08/project8/Parser.py

class Parser:
    def __init__(self, input_file_path):
        """
        Opens the input file/stream, and gets ready to parse it
        Arguments: Input file path
        """
        self.file = open(input_file_path, 'r')
        self.current_line = ''

    def hasMoreLines(self) -> bool:
        """
        Are there more lines in the input?
        """
        self.current_line = self.file.readline()
        return self.current_line != ''

    def advance(self):
        """
        Reads the next command from the input and makes it the current command.
        This method should be called if hasMoreLines is true.
        Initially there is no current command
        """
        # Assuming current_line already contains the next line from has_more_lines
        # Trim the current line to remove leading and trailing whitespace
        self.current_line = self.current_line.strip()

        # If the line is a comment or empty, keep reading the next line
        while self.current_line == '' or self.current_line.startswith('/'):
            self.current_line = self.file.readline()
            if self.current_line == '':  # End of file reached
                break
            self.current_line = self.current_line.strip()

    def commandType(self) -> str:
        """
        Returns a constant representing the type of the current command.
        If the current command is an arithmetic-logical command, return C_ARITHMETIC.
        Return options: C_ARITHMETIC, C_PUSH, C_POP, C_LABEL, C_GOTO, C_IF, C_FUNCTION, C_RETURN, C_CALL
        """
        arith = ['add', 'sub', 'neg', 'eq', 'gt', 'lt', 'and', 'or', 'not']
        command = str(self.current_line).split()[0]

        if command in arith: return "C_ARITHMETIC"
        elif command == "push": return "C_PUSH"
        elif command == "pop": return "C_POP"
        elif command == "label": return "C_LABEL"
        elif command == "goto": return "C_GOTO"
        elif command == "if-goto": return "C_IF"
        elif command == "function": return "C_FUNCTION"
        elif command == "return": return "C_RETURN"
        elif command == "call": return "C_CALL"
        else: return "something Wrong ! ! ! !"

    def arg1(self) -> str:
        """
        Return the first argument of the current command.
        In the case of C_ARITHMETIC, the command itself(add, sub...) is returned.
        Should not be called if the current command is C_RETURN
        """
        arith = ['add', 'sub', 'neg', 'eq', 'gt', 'lt', 'and', 'or', 'not']
        command = str(self.current_line).split()[0]
        if command in arith: return command
        else: return str(self.current_line).split()[1]

    def arg2(self) -> int:
        """
        Return the second argument of the current command.
        Should be called only if the current command is C_PUSH, C_POP, C_FUNCTION, C_CALL
        """
        return self.current_line.split()[2]

    def close(self):
        """
        Closes the file stream.
        """
        self.file.close()